File: src/lib/schema-lint/validate.ts
import {
  COURSE_FAMILY,
  LOCAL_BUSINESS_FAMILY,
  familyFor,
  matchesRequested,
  typesOf,
} from "./const";
import { extractJsonLd } from "./extract";
import { asList, hasText, nonempty, textOf } from "./props";
import {
  DEFAULT_TYPES,
  OPTIONAL_TYPES,
  type CheckOptions,
  type Entity,
  type Extraction,
  type Issue,
  type UrlResult,
} from "./types";

function err(type: string, path: string, message: string): Issue {
  return { level: "error", type, path, message };
}
function warn(type: string, path: string, message: string): Issue {
  return { level: "warning", type, path, message };
}

function schemaOrgContext(ctx: unknown): boolean {
  if (ctx == null) return false;
  if (typeof ctx === "string") return ctx.toLowerCase().includes("schema.org");
  if (Array.isArray(ctx)) return ctx.some(schemaOrgContext);
  if (typeof ctx === "object") {
    const rec = ctx as Record<string, unknown>;
    if (schemaOrgContext(rec["@vocab"])) return true;
    return Object.values(rec).some(schemaOrgContext);
  }
  return false;
}

function primaryType(types: string[]): string {
  for (const t of types) {
    if (LOCAL_BUSINESS_FAMILY.has(t) || COURSE_FAMILY.has(t) || t === "FAQPage" || OPTIONAL_TYPES.includes(t as (typeof OPTIONAL_TYPES)[number])) {
      return t;
    }
  }
  return types[0] ?? "";
}

function inFilter(found: string[], requested: string[] | null | undefined): boolean {
  if (requested == null) return matchesRequested(found, [...DEFAULT_TYPES]);
  if (!requested.length) return true;
  return matchesRequested(found, requested);
}

function optionalOn(requested: string[] | null | undefined): boolean {
  if (requested == null || !requested.length) return true;
  return OPTIONAL_TYPES.some((opt) => inFilter([opt], requested));
}

function isLocal(types: string[]): boolean {
  return types.some((t) => LOCAL_BUSINESS_FAMILY.has(t));
}
function isCourse(types: string[]): boolean {
  return types.some((t) => COURSE_FAMILY.has(t));
}

function validateCourseInstance(node: Record<string, unknown>, issues: Issue[], path: string, label = "CourseInstance") {
  if (!hasText(node, "startDate")) {
    issues.push(warn(label, `${path}.startDate`, "CourseInstance missing startDate"));
  }
  if (!hasText(node, "endDate") && !hasText(node, "courseMode")) {
    issues.push(warn(label, `${path}.endDate`, "CourseInstance missing endDate or courseMode"));
  }
}

function validateCourse(ent: Entity, issues: Issue[]) {
  const node = ent.node;
  const label = primaryType(ent.types) || "Course";
  const base = ent.path;
  if (ent.types.includes("Course") || (!ent.types.includes("CourseInstance") && isCourse(ent.types))) {
    if (!hasText(node, "name")) issues.push(err(label, `${base}.name`, "missing required property name"));
    if (!hasText(node, "description") && !nonempty(node.about)) {
      issues.push(err(label, `${base}.description`, "missing required property description (or about)"));
    }
    if (!["provider", "organizer", "author"].some((k) => nonempty(node[k]))) {
      issues.push(err(label, `${base}.provider`, "missing required provider, organizer, or author"));
    }
    if (!hasText(node, "url")) issues.push(warn(label, `${base}.url`, "recommended property url is missing"));
    const offers = node.offers;
    if (!nonempty(offers)) {
      issues.push(warn(label, `${base}.offers`, "recommended property offers is missing"));
    } else {
      let hasPrice = false;
      for (const off of asList(offers)) {
        if (off && typeof off === "object" && !Array.isArray(off)) {
          const rec = off as Record<string, unknown>;
          if (hasText(rec, "price") || hasText(rec, "lowPrice") || hasText(rec, "highPrice")) {
            hasPrice = true;
            break;
          }
        }
        if (textOf(off)) {
          hasPrice = true;
          break;
        }
      }
      if (!hasPrice) issues.push(warn(label, `${base}.offers.price`, "recommended offers.price is missing"));
    }
    const instances = asList(node.hasCourseInstance);
    if (!instances.length) {
      issues.push(warn(label, `${base}.hasCourseInstance`, "recommended property hasCourseInstance is missing"));
    } else {
      instances.forEach((inst, i) => {
        if (inst && typeof inst === "object" && !Array.isArray(inst)) {
          validateCourseInstance(inst as Record<string, unknown>, issues, `${base}.hasCourseInstance[${i}]`);
        } else {
          issues.push(warn(label, `${base}.hasCourseInstance[${i}]`, "CourseInstance is not an object; cannot check dates"));
        }
      });
    }
  }
  if (ent.types.includes("CourseInstance")) {
    validateCourseInstance(node, issues, base, "CourseInstance");
  }
}

function addressOk(node: Record<string, unknown>): boolean {
  const addr = node.address;
  if (addr == null) return false;
  if (typeof addr === "string" && addr.trim()) return true;
  for (const item of asList(addr)) {
    if (typeof item === "string" && item.trim()) return true;
    if (item && typeof item === "object" && !Array.isArray(item)) {
      const rec = item as Record<string, unknown>;
      if (textOf(rec.streetAddress) || textOf(rec.addressLocality)) return true;
    }
  }
  return false;
}

function validateLocal(ent: Entity, issues: Issue[]) {
  const node = ent.node;
  const label = primaryType(ent.types) || "LocalBusiness";
  const base = ent.path;
  if (!hasText(node, "name")) issues.push(err(label, `${base}.name`, "missing required property name"));
  const hasMap = nonempty(node.hasMap);
  if (!addressOk(node) && !hasMap) {
    issues.push(
      err(
        label,
        `${base}.address`,
        "missing required address (PostalAddress with streetAddress or addressLocality) or hasMap",
      ),
    );
  } else if (!addressOk(node) && hasMap) {
    issues.push(warn(label, `${base}.address`, "address missing; hasMap is present so location requirement is met"));
  }
  if (!hasText(node, "telephone") && !hasText(node, "url")) {
    issues.push(err(label, `${base}.telephone`, "missing telephone and url (at least one is required)"));
  }
  if (!nonempty(node.openingHoursSpecification) && !hasText(node, "openingHours")) {
    issues.push(warn(label, `${base}.openingHours`, "recommended openingHours or openingHoursSpecification is missing"));
  }
  if (!nonempty(node.geo)) issues.push(warn(label, `${base}.geo`, "recommended property geo is missing"));
  if (!nonempty(node.image)) issues.push(warn(label, `${base}.image`, "recommended property image is missing"));
  if (!hasText(node, "priceRange")) issues.push(warn(label, `${base}.priceRange`, "recommended property priceRange is missing"));
}

function answerText(q: Record<string, unknown>): boolean {
  const ans = q.acceptedAnswer;
  if (!nonempty(ans)) return false;
  for (const item of asList(ans)) {
    if (typeof item === "string" && item.trim()) return true;
    if (item && typeof item === "object" && !Array.isArray(item) && hasText(item as Record<string, unknown>, "text", "name")) {
      return true;
    }
  }
  return false;
}

function validateFaq(ent: Entity, issues: Issue[]) {
  const node = ent.node;
  const label = "FAQPage";
  const base = ent.path;
  const main = node.mainEntity;
  if (main == null) {
    issues.push(err(label, `${base}.mainEntity`, "missing required property mainEntity"));
    return;
  }
  const questions = asList(main);
  if (!questions.length) {
    issues.push(err(label, `${base}.mainEntity`, "FAQPage present with empty mainEntity"));
    return;
  }
  const expanded: Array<[string, unknown]> = [];
  questions.forEach((q, i) => {
    if (q && typeof q === "object" && !Array.isArray(q) && typesOf(q).join() === "ItemList") {
      asList((q as Record<string, unknown>).itemListElement).forEach((el, j) => {
        const item =
          el && typeof el === "object" && !Array.isArray(el) && "item" in el
            ? (el as Record<string, unknown>).item
            : el;
        expanded.push([`${base}.mainEntity[${i}].itemListElement[${j}]`, item]);
      });
    } else {
      expanded.push([`${base}.mainEntity[${i}]`, q]);
    }
  });
  if (!expanded.length) {
    issues.push(err(label, `${base}.mainEntity`, "FAQPage present with empty mainEntity"));
    return;
  }
  for (const [path, q] of expanded) {
    if (!q || typeof q !== "object" || Array.isArray(q)) {
      issues.push(err(label, path, "mainEntity item is not a Question object"));
      continue;
    }
    const rec = q as Record<string, unknown>;
    const qtypes = typesOf(rec);
    if (qtypes.length && !qtypes.includes("Question")) {
      issues.push(err(label, `${path}.@type`, `mainEntity item @type is ${qtypes[0]}, expected Question`));
    }
    if (!hasText(rec, "name", "text")) issues.push(err(label, `${path}.name`, "Question missing name or text"));
    if (!nonempty(rec.acceptedAnswer)) {
      issues.push(err(label, `${path}.acceptedAnswer`, "Question missing acceptedAnswer"));
    } else if (!answerText(rec)) {
      issues.push(err(label, `${path}.acceptedAnswer.text`, "acceptedAnswer missing text"));
    }
  }
}

function validateOrg(ent: Entity, issues: Issue[]) {
  const node = ent.node;
  const label = primaryType(ent.types) || "Organization";
  if (!hasText(node, "name")) issues.push(warn(label, `${ent.path}.name`, "Organization missing name"));
  if (!hasText(node, "url")) issues.push(warn(label, `${ent.path}.url`, "Organization missing url"));
}

function validateWebsite(ent: Entity, issues: Issue[]) {
  if (!hasText(ent.node, "name")) issues.push(warn("WebSite", `${ent.path}.name`, "WebSite missing name"));
  if (!hasText(ent.node, "url")) issues.push(warn("WebSite", `${ent.path}.url`, "WebSite missing url"));
}

function validateBreadcrumb(ent: Entity, issues: Issue[]) {
  const elements = asList(ent.node.itemListElement);
  if (!elements.length) {
    issues.push(warn("BreadcrumbList", `${ent.path}.itemListElement`, "BreadcrumbList missing itemListElement"));
    return;
  }
  elements.forEach((el, i) => {
    const path = `${ent.path}.itemListElement[${i}]`;
    if (el && typeof el === "object" && !Array.isArray(el)) {
      const rec = el as Record<string, unknown>;
      const hasName =
        hasText(rec, "name") ||
        (rec.item && typeof rec.item === "object" && !Array.isArray(rec.item) && hasText(rec.item as Record<string, unknown>, "name")) ||
        (typeof rec.item === "string" && rec.item.trim().length > 0);
      if (!hasName) issues.push(warn("BreadcrumbList", `${path}.name`, "ListItem missing name"));
      if ("item" in rec || "position" in rec || "name" in rec) {
        if (!hasText(rec, "position")) issues.push(warn("BreadcrumbList", `${path}.position`, "ListItem missing position"));
      }
    } else {
      issues.push(warn("BreadcrumbList", path, "breadcrumb entry is not an object"));
    }
  });
}

export function validateEntities(entities: Entity[], requestedTypes: string[] | null = null): Issue[] {
  const issues: Issue[] = [];
  const opt = optionalOn(requestedTypes);
  for (const ent of entities) {
    const types = ent.types;
    if (!types.length) continue;
    const runCourse = isCourse(types) && inFilter(["Course"], requestedTypes);
    const runLocal = isLocal(types) && inFilter(["LocalBusiness"], requestedTypes);
    const runFaq = types.includes("FAQPage") && inFilter(["FAQPage"], requestedTypes);
    const runOrg =
      opt &&
      !isLocal(types) &&
      (types.includes("Organization") || types.some((t) => familyFor("Organization").has(t))) &&
      inFilter(["Organization"], requestedTypes ?? ["Organization"]);
    const runSite = opt && types.includes("WebSite") && (requestedTypes == null || inFilter(["WebSite"], requestedTypes));
    const runCrumbs =
      opt && types.includes("BreadcrumbList") && (requestedTypes == null || inFilter(["BreadcrumbList"], requestedTypes));
    const active = runCourse || runLocal || runFaq || runOrg || runSite || runCrumbs;
    if (active && !schemaOrgContext(ent.context)) {
      issues.push(warn(primaryType(types) || "JSON-LD", `${ent.path}.@context`, "missing schema.org @context"));
    }
    if (runCourse) validateCourse(ent, issues);
    if (runLocal) validateLocal(ent, issues);
    if (runFaq) validateFaq(ent, issues);
    if (runOrg) validateOrg(ent, issues);
    if (runSite) validateWebsite(ent, issues);
    if (runCrumbs) validateBreadcrumb(ent, issues);
  }
  return issues;
}

export function checkExtraction(
  extraction: Extraction,
  opts: { url: string; requestedTypes?: string[] | null; strictTypes?: boolean; failOnFetch?: boolean; fetchError?: string | null },
): UrlResult {
  if (opts.fetchError) {
    return {
      url: opts.url,
      status: "FETCH_FAILED",
      types_found: [],
      issues: [
        {
          level: opts.failOnFetch === false ? "warning" : "error",
          type: "",
          path: "",
          message: opts.fetchError,
        },
      ],
      jsonld_count: 0,
    };
  }
  const issues: Issue[] = [];
  for (const block of extraction.blocks) {
    if (block.error) issues.push(err("JSON-LD", `$blocks[${block.index}]`, block.error));
  }
  issues.push(...validateEntities(extraction.entities, opts.requestedTypes ?? null));
  const strictAgainst = opts.requestedTypes == null ? [...DEFAULT_TYPES] : opts.requestedTypes;
  if (opts.strictTypes && !matchesRequested(extraction.typesFound, strictAgainst)) {
    const found = extraction.typesFound.length ? extraction.typesFound.join(", ") : "(none)";
    const want = strictAgainst.length ? strictAgainst.join(", ") : "(any)";
    issues.push(err("", "$", `none of the requested types present: ${want} (found: ${found})`));
  }
  const errors = issues.filter((i) => i.level === "error");
  return {
    url: opts.url,
    status: errors.length ? "ERRORS" : "OK",
    types_found: [...extraction.typesFound],
    issues,
    jsonld_count: extraction.blocks.length,
  };
}

export function checkHtml(html: string, url = "(html)", options: CheckOptions = {}): UrlResult {
  const extraction = extractJsonLd(html);
  return checkExtraction(extraction, {
    url,
    requestedTypes: options.types ?? null,
    strictTypes: options.strictTypes ?? false,
    failOnFetch: options.failOnFetch ?? true,
  });
}

export function buildReport(results: UrlResult[]): {
  schema: "schema-lint/v1";
  generated_at: string;
  results: UrlResult[];
  error_count: number;
  warning_count: number;
} {
  const error_count = results.reduce((n, r) => n + r.issues.filter((i) => i.level === "error").length, 0);
  const warning_count = results.reduce((n, r) => n + r.issues.filter((i) => i.level === "warning").length, 0);
  return {
    schema: "schema-lint/v1",
    generated_at: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    results,
    error_count,
    warning_count,
  };
}

File: schema_lint/validate.py
"""Validate extracted entities against documented required properties.

Missing required props = errors. Never invent fields as present.
"""

from __future__ import annotations

from typing import Any, Iterable

from schema_lint.const import (
    COURSE_FAMILY,
    DEFAULT_TYPES,
    LOCAL_BUSINESS_FAMILY,
    OPTIONAL_TYPES,
    family_for,
    is_course_family,
    is_local_business,
    matches_requested,
    types_of,
)
from schema_lint.extract import Entity, Extraction, extract_jsonld
from schema_lint.fetch import FetchResult, fetch
from schema_lint.models import Issue, UrlResult, issue_error, issue_warn
from schema_lint.props import as_list, has_text, nonempty, text_of


def parse_types_arg(types: str | list[str] | None) -> list[str] | None:
    """None means CLI/library default. Empty list means 'no type filter'."""
    if types is None:
        return None
    if isinstance(types, list):
        return [t.strip() for t in types if str(t).strip()]
    parts = [p.strip() for p in str(types).split(",") if p.strip()]
    return parts


def _schema_org_context(ctx: Any) -> bool:
    if ctx is None:
        return False
    if isinstance(ctx, str):
        return "schema.org" in ctx.lower()
    if isinstance(ctx, list):
        return any(_schema_org_context(x) for x in ctx)
    if isinstance(ctx, dict):
        if _schema_org_context(ctx.get("@vocab")):
            return True
        return any(_schema_org_context(v) for v in ctx.values())
    return False


def _primary_type(types: list[str]) -> str:
    for t in types:
        if t in LOCAL_BUSINESS_FAMILY:
            return t
        if t in COURSE_FAMILY:
            return t
        if t == "FAQPage":
            return t
        if t in OPTIONAL_TYPES:
            return t
    return types[0] if types else ""


def _in_filter(found: list[str], requested: list[str] | None) -> bool:
    if requested is None:
        # default: primary families
        return matches_requested(found, list(DEFAULT_TYPES))
    if not requested:
        return True
    return matches_requested(found, requested)


def _optional_enabled(requested: list[str] | None) -> bool:
    """Optional types are warn-only and on by default; filtered out if --types omits them."""
    if requested is None:
        return True
    if not requested:
        return True
    return any(_in_filter([opt], requested) for opt in OPTIONAL_TYPES)


def _validate_course(ent: Entity, issues: list[Issue]) -> None:
    node = ent.node
    label = _primary_type(ent.types) or "Course"
    base = ent.path

    if "Course" in ent.types or ("CourseInstance" not in ent.types and is_course_family(ent.types)):
        if not has_text(node, "name"):
            issues.append(issue_error(label, f"{base}.name", "missing required property name"))
        if not has_text(node, "description") and not nonempty(node.get("about")):
            issues.append(
                issue_error(
                    label,
                    f"{base}.description",
                    "missing required property description (or about)",
                )
            )
        if not any(nonempty(node.get(k)) for k in ("provider", "organizer", "author")):
            issues.append(
                issue_error(
                    label,
                    f"{base}.provider",
                    "missing required provider, organizer, or author",
                )
            )
        if not has_text(node, "url"):
            issues.append(issue_warn(label, f"{base}.url", "recommended property url is missing"))
        offers = node.get("offers")
        if not nonempty(offers):
            issues.append(issue_warn(label, f"{base}.offers", "recommended property offers is missing"))
        else:
            has_price = False
            for off in as_list(offers):
                if isinstance(off, dict) and (
                    has_text(off, "price") or has_text(off, "lowPrice") or has_text(off, "highPrice")
                ):
                    has_price = True
                    break
                if text_of(off):
                    has_price = True
            if not has_price:
                issues.append(
                    issue_warn(label, f"{base}.offers.price", "recommended offers.price is missing")
                )
        instances = as_list(node.get("hasCourseInstance"))
        if not instances:
            issues.append(
                issue_warn(
                    label,
                    f"{base}.hasCourseInstance",
                    "recommended property hasCourseInstance is missing",
                )
            )
        else:
            for i, inst in enumerate(instances):
                if isinstance(inst, dict):
                    _validate_course_instance(inst, issues, f"{base}.hasCourseInstance[{i}]")
                else:
                    issues.append(
                        issue_warn(
                            label,
                            f"{base}.hasCourseInstance[{i}]",
                            "CourseInstance is not an object; cannot check dates",
                        )
                    )

    if "CourseInstance" in ent.types:
        _validate_course_instance(node, issues, base, label="CourseInstance")


def _validate_course_instance(
    node: dict[str, Any],
    issues: list[Issue],
    path: str,
    *,
    label: str = "CourseInstance",
) -> None:
    if not has_text(node, "startDate"):
        issues.append(issue_warn(label, f"{path}.startDate", "CourseInstance missing startDate"))
    if not has_text(node, "endDate") and not has_text(node, "courseMode"):
        issues.append(
            issue_warn(label, f"{path}.endDate", "CourseInstance missing endDate or courseMode")
        )


def _address_ok(node: dict[str, Any]) -> bool:
    addr = node.get("address")
    if addr is None:
        return False
    if isinstance(addr, str) and addr.strip():
        return True
    for item in as_list(addr):
        if isinstance(item, str) and item.strip():
            return True
        if isinstance(item, dict):
            if text_of(item.get("streetAddress")) or text_of(item.get("addressLocality")):
                return True
    return False


def _has_map(node: dict[str, Any]) -> bool:
    return nonempty(node.get("hasMap"))


def _validate_local_business(ent: Entity, issues: list[Issue]) -> None:
    node = ent.node
    label = _primary_type(ent.types) or "LocalBusiness"
    base = ent.path

    if not has_text(node, "name"):
        issues.append(issue_error(label, f"{base}.name", "missing required property name"))

    if not _address_ok(node) and not _has_map(node):
        issues.append(
            issue_error(
                label,
                f"{base}.address",
                "missing required address (PostalAddress with streetAddress or addressLocality) or hasMap",
            )
        )
    elif not _address_ok(node) and _has_map(node):
        issues.append(
            issue_warn(
                label,
                f"{base}.address",
                "address missing; hasMap is present so location requirement is met",
            )
        )

    if not has_text(node, "telephone") and not has_text(node, "url"):
        issues.append(
            issue_error(
                label,
                f"{base}.telephone",
                "missing telephone and url (at least one is required)",
            )
        )

    if not nonempty(node.get("openingHoursSpecification")) and not has_text(node, "openingHours"):
        issues.append(
            issue_warn(
                label,
                f"{base}.openingHours",
                "recommended openingHours or openingHoursSpecification is missing",
            )
        )
    if not nonempty(node.get("geo")):
        issues.append(issue_warn(label, f"{base}.geo", "recommended property geo is missing"))
    if not nonempty(node.get("image")):
        issues.append(issue_warn(label, f"{base}.image", "recommended property image is missing"))
    if not has_text(node, "priceRange"):
        issues.append(issue_warn(label, f"{base}.priceRange", "recommended property priceRange is missing"))


def _answer_text(q: dict[str, Any]) -> bool:
    ans = q.get("acceptedAnswer")
    if not nonempty(ans):
        return False
    for item in as_list(ans):
        if isinstance(item, str) and item.strip():
            return True
        if isinstance(item, dict) and has_text(item, "text", "name"):
            return True
    return False


def _validate_faq(ent: Entity, issues: list[Issue]) -> None:
    node = ent.node
    label = "FAQPage"
    base = ent.path
    main = node.get("mainEntity")
    if main is None:
        issues.append(issue_error(label, f"{base}.mainEntity", "missing required property mainEntity"))
        return
    questions = as_list(main)
    if len(questions) == 0:
        issues.append(issue_error(label, f"{base}.mainEntity", "FAQPage present with empty mainEntity"))
        return

    expanded: list[tuple[str, Any]] = []
    for i, q in enumerate(questions):
        if isinstance(q, dict) and "Question" not in types_of(q) and types_of(q) == ["ItemList"]:
            for j, el in enumerate(as_list(q.get("itemListElement"))):
                item = el.get("item") if isinstance(el, dict) and "item" in el else el
                expanded.append((f"{base}.mainEntity[{i}].itemListElement[{j}]", item))
        else:
            expanded.append((f"{base}.mainEntity[{i}]", q))

    if not expanded:
        issues.append(issue_error(label, f"{base}.mainEntity", "FAQPage present with empty mainEntity"))
        return

    for path, q in expanded:
        if not isinstance(q, dict):
            issues.append(issue_error(label, path, "mainEntity item is not a Question object"))
            continue
        qtypes = types_of(q)
        if qtypes and "Question" not in qtypes:
            issues.append(
                issue_error(label, f"{path}.@type", f"mainEntity item @type is {qtypes[0]}, expected Question")
            )
        if not has_text(q, "name", "text"):
            issues.append(issue_error(label, f"{path}.name", "Question missing name or text"))
        if not nonempty(q.get("acceptedAnswer")):
            issues.append(issue_error(label, f"{path}.acceptedAnswer", "Question missing acceptedAnswer"))
        elif not _answer_text(q):
            issues.append(
                issue_error(label, f"{path}.acceptedAnswer.text", "acceptedAnswer missing text")
            )


def _validate_organization(ent: Entity, issues: list[Issue]) -> None:
    node = ent.node
    label = _primary_type(ent.types) or "Organization"
    base = ent.path
    if not has_text(node, "name"):
        issues.append(issue_warn(label, f"{base}.name", "Organization missing name"))
    if not has_text(node, "url"):
        issues.append(issue_warn(label, f"{base}.url", "Organization missing url"))


def _validate_website(ent: Entity, issues: list[Issue]) -> None:
    node = ent.node
    base = ent.path
    if not has_text(node, "name"):
        issues.append(issue_warn("WebSite", f"{base}.name", "WebSite missing name"))
    if not has_text(node, "url"):
        issues.append(issue_warn("WebSite", f"{base}.url", "WebSite missing url"))


def _validate_breadcrumb(ent: Entity, issues: list[Issue]) -> None:
    node = ent.node
    base = ent.path
    elements = as_list(node.get("itemListElement"))
    if not elements:
        issues.append(
            issue_warn("BreadcrumbList", f"{base}.itemListElement", "BreadcrumbList missing itemListElement")
        )
        return
    for i, el in enumerate(elements):
        path = f"{base}.itemListElement[{i}]"
        if isinstance(el, dict) and ("item" in el or "position" in el or "name" in el):
            has_name = has_text(el, "name") or (
                isinstance(el.get("item"), dict) and has_text(el["item"], "name")
            ) or (isinstance(el.get("item"), str) and bool(el["item"].strip()))
            if not has_name:
                issues.append(issue_warn("BreadcrumbList", f"{path}.name", "ListItem missing name"))
            if not has_text(el, "position"):
                issues.append(issue_warn("BreadcrumbList", f"{path}.position", "ListItem missing position"))
        elif isinstance(el, dict):
            if not has_text(el, "name"):
                issues.append(issue_warn("BreadcrumbList", f"{path}.name", "breadcrumb entry missing name"))
        else:
            issues.append(issue_warn("BreadcrumbList", path, "breadcrumb entry is not an object"))


def validate_entities(
    entities: Iterable[Entity],
    *,
    requested_types: list[str] | None = None,
) -> list[Issue]:
    issues: list[Issue] = []
    optional_on = _optional_enabled(requested_types)

    for ent in entities:
        types = ent.types
        if not types:
            continue

        run_course = is_course_family(types) and _in_filter(["Course"], requested_types)
        run_local = is_local_business(types) and _in_filter(["LocalBusiness"], requested_types)
        run_faq = "FAQPage" in types and _in_filter(["FAQPage"], requested_types)
        run_org = (
            optional_on
            and not is_local_business(types)
            and (any(t == "Organization" for t in types) or any(t in family_for("Organization") for t in types))
            and _in_filter(["Organization"], requested_types if requested_types is not None else ["Organization"])
        )
        run_site = optional_on and "WebSite" in types and (
            requested_types is None or _in_filter(["WebSite"], requested_types)
        )
        run_crumbs = optional_on and "BreadcrumbList" in types and (
            requested_types is None or _in_filter(["BreadcrumbList"], requested_types)
        )

        active = run_course or run_local or run_faq or run_org or run_site or run_crumbs
        if active and not _schema_org_context(ent.context):
            issues.append(
                issue_warn(
                    _primary_type(types) or "JSON-LD",
                    f"{ent.path}.@context",
                    "missing schema.org @context",
                )
            )

        if run_course:
            _validate_course(ent, issues)
        if run_local:
            _validate_local_business(ent, issues)
        if run_faq:
            _validate_faq(ent, issues)
        if run_org:
            _validate_organization(ent, issues)
        if run_site:
            _validate_website(ent, issues)
        if run_crumbs:
            _validate_breadcrumb(ent, issues)

    return issues


def check_extraction(
    extraction: Extraction,
    *,
    url: str,
    requested_types: list[str] | None = None,
    strict_types: bool = False,
    fail_on_fetch: bool = True,
    fetch_error: str | None = None,
) -> UrlResult:
    if fetch_error:
        level = "error" if fail_on_fetch else "warning"
        return UrlResult(
            url=url,
            status="FETCH_FAILED",
            types_found=[],
            issues=[Issue(level=level, type="", path="", message=fetch_error)],
            jsonld_count=0,
        )

    issues: list[Issue] = []
    for block in extraction.blocks:
        if block.error:
            issues.append(issue_error("JSON-LD", f"$blocks[{block.index}]", block.error))

    issues.extend(validate_entities(extraction.entities, requested_types=requested_types))

    strict_against = list(DEFAULT_TYPES) if requested_types is None else list(requested_types)
    if strict_types:
        if not matches_requested(extraction.types_found, strict_against):
            found = ", ".join(extraction.types_found) if extraction.types_found else "(none)"
            want = ", ".join(strict_against) if strict_against else "(any)"
            issues.append(
                issue_error(
                    "",
                    "$",
                    f"none of the requested types present: {want} (found: {found})",
                )
            )

    errors = [i for i in issues if i.level == "error"]
    status = "ERRORS" if errors else "OK"
    return UrlResult(
        url=url,
        status=status,
        types_found=list(extraction.types_found),
        issues=issues,
        jsonld_count=len(extraction.blocks),
    )


def check_html(
    html: str,
    *,
    url: str = "(html)",
    types: str | list[str] | None = None,
    strict_types: bool = False,
    fail_on_fetch: bool = True,
) -> UrlResult:
    requested = parse_types_arg(types)
    extraction = extract_jsonld(html)
    return check_extraction(
        extraction,
        url=url,
        requested_types=requested,
        strict_types=strict_types,
        fail_on_fetch=fail_on_fetch,
    )


def check_url(
    target: str,
    *,
    types: str | list[str] | None = None,
    strict_types: bool = False,
    timeout_sec: int = 30,
    user_agent: str | None = None,
    fail_on_fetch: bool = True,
) -> tuple[UrlResult, FetchResult]:
    from schema_lint import DEFAULT_USER_AGENT

    fetched = fetch(target, timeout_sec=timeout_sec, user_agent=user_agent or DEFAULT_USER_AGENT)
    requested = parse_types_arg(types)
    if not fetched.ok:
        result = check_extraction(
            Extraction(),
            url=target,
            requested_types=requested,
            strict_types=strict_types,
            fail_on_fetch=fail_on_fetch,
            fetch_error=fetched.error or "fetch failed",
        )
        return result, fetched
    result = check_html(
        fetched.html,
        url=target,
        types=requested,
        strict_types=strict_types,
        fail_on_fetch=fail_on_fetch,
    )
    return result, fetched
